[SVG] Add fast mode to repaintRectInLocalCoordinates

Constellation opened the pull request:
#### b21509bc0fbc78ae8f1684d7211d041fbfb92038
<pre>
[SVG] Add fast mode to repaintRectInLocalCoordinates
<a href="https://bugs.webkit.org/show_bug.cgi?id=262672">https://bugs.webkit.org/show_bug.cgi?id=262672</a>
rdar://116499404

Reviewed by Said Abou-Hallawa.

This patch replaces repaintRectInLocalCoordinates calls with fast approximation version stubs. They are no difference at this point,
but approximation version will get implementation in a subsequent change.

1. This patch adds Fast and Accurate mode to repaintRectInLocalCoordinates. And replaces fastRepaintRectInLocalCoordinates with repaintRectInLocalCoordinates Fast mode.
2. We use repaintRectInLocalCoordinates Accurate mode in checkEnclosure / checkIntersection to keep the current behavior as is even after introducing approximation.
2. We add visibleRectContextForRenderTreeAsText to continue using repaintRectInLocalCoordinates Accurate mode for SVG RenderTreeAsText. This is having an option
   in VisibleRectContext and switch the calls based on this option.

* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::absoluteClippedOverflowRectForRenderTreeAsText const):
(WebCore::RenderObject::visibleRectContextForRenderTreeAsText):
* Source/WebCore/rendering/svg/RenderSVGBlock.cpp:
(WebCore::RenderSVGBlock::clippedOverflowRect const):
* Source/WebCore/rendering/svg/RenderSVGInline.cpp:
(WebCore::RenderSVGInline::fastRepaintRectInLocalCoordinates const):
(WebCore::RenderSVGInline::clippedOverflowRect const):
* Source/WebCore/rendering/svg/RenderSVGInline.h:
* Source/WebCore/rendering/svg/RenderSVGModelObject.cpp:
(WebCore::RenderSVGModelObject::addFocusRingRects const):
(WebCore::RenderSVGModelObject::checkIntersection):
(WebCore::RenderSVGModelObject::checkEnclosure):
* Source/WebCore/rendering/svg/RenderSVGResourceGradient.cpp:
(WebCore::createMaskAndSwapContextForTextGradient):
(WebCore::clipToTextMask):
* Source/WebCore/rendering/svg/RenderSVGResourceMarker.cpp:
(WebCore::RenderSVGResourceMarker::markerBoundaries const):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
(WebCore::RenderSVGResourceMasker::applyResource):
(WebCore::RenderSVGResourceMasker::calculateMaskContentRepaintRect):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::clippedOverflowRectForRepaint):
(WebCore::SVGRenderSupport::computeContainerBoundingBoxes):
* Source/WebCore/rendering/svg/SVGRenderSupport.h:
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
* Source/WebCore/rendering/svg/SVGRenderingContext.cpp:
(WebCore::SVGRenderingContext::prepareToRenderSVGContent):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.cpp:
(WebCore::LegacyRenderSVGContainer::addFocusRingRects const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.cpp:
(WebCore::LegacyRenderSVGImage::paint):
(WebCore::LegacyRenderSVGImage::addFocusRingRects const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.cpp:
(WebCore::LegacyRenderSVGModelObject::clippedOverflowRect const):
(WebCore::LegacyRenderSVGModelObject::outlineBoundsForRepaint const):
(WebCore::LegacyRenderSVGModelObject::absoluteFocusRingQuads):
(WebCore::LegacyRenderSVGModelObject::checkIntersection):
(WebCore::LegacyRenderSVGModelObject::checkEnclosure):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.cpp:
(WebCore::LegacyRenderSVGResourceClipper::applyResource):
(WebCore::LegacyRenderSVGResourceClipper::calculateClipContentRepaintRect):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::layout):
(WebCore::LegacyRenderSVGRoot::clippedOverflowRect const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::paint):
(WebCore::LegacyRenderSVGShape::addFocusRingRects const):
* Source/WebCore/svg/SVGFEImageElement.cpp:
(WebCore::SVGFEImageElement::imageBufferForEffect const):

Canonical link: <a href="https://commits.webkit.org/268940@main">https://commits.webkit.org/268940@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/be8f23b31deb942c45ff6ea9b57c6b7c6f8da8af

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21114 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/21485 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22178 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22992 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/19637 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21354 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24744 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21676 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20878 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21338 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21054 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/18312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23846 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/18212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/19142 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/25411 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19293 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/19343 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23346 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/19884 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/16903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/19163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/23452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2612 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19738 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -789,6 +790,7 @@ class RenderObject : public CachedImageClient, public CanMakeCheckedPtr {
         ApplyCompositedClips = 1 << 1,
         ApplyCompositedContainerScrolls  = 1 << 2,
         ApplyContainerClip = 1 << 3,
+        DoNotUseSVGFastRepaintRectInLocalCoordinates = 1 << 4,
```
I think this is a long name. And using the negative imperative makes it harder to read. Can't we use something like `UseFastRepaintRect` or `CalculateFastRepaintRect` or something similar? All the renderers expect the SVG ones are already doing fast repaintRect calculation. So I think it makes sense to be added by default except where we do need accurate repaintRect from SVG renderers.

Constellation replied:
Changed to `CalculateFastRepaintRect`

shallawa commented on Source/WebCore/rendering/svg/RenderSVGModelObject.cpp:
```diff
@@ -224,6 +224,7 @@ bool RenderSVGModelObject::checkIntersection(RenderElement* renderer, const Floa
     SVGElement* svgElement = downcast<SVGElement>(renderer->element());
     ASSERT(is<SVGGraphicsElement>(svgElement));
     auto ctm = downcast<SVGGraphicsElement>(*svgElement).getCTM(SVGLocatable::DisallowStyleUpdate);
+    // FIXME: Using repaintRect in hit-detection is wrong.
```
I this comment is misleading. I think the calculation of r`epaintRectInLocalCoordinates()` is correct and using it here is correct. The naming might be the problem. So I think this comment might be added in RenderObject.h where the function is defined.

Also FIXME comments should not be about bugs in the code. If you think `repaintRectInLocalCoordinates()` has bugs or this caller should use different rectangle, please file a bug instead.

Constellation replied:
I don't think using `repaintRectInLocalCoordinates` is correct. This is already inconsistent with the implementation of hit-testing in SVG renderer, they are not accounting filter and masker. So, this is broken.

Constellation replied:
Anyway, I'll file a bug and put URL to this FIXME. I think both naming and implementation are completely wrong and broken.

Constellation replied:
Filed and added URL to this FIXME. https://bugs.webkit.org/show_bug.cgi?id=262709
This implementation is completely broken.

shallawa commented on Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
```diff
@@ -55,16 +55,20 @@
+    if (context.options.contains(RenderObject::VisibleRectContextOption::DoNotUseSVGFastRepaintRectInLocalCoordinates))
+        repaintRect = renderer.repaintRectInLocalCoordinates();
+    else
+        repaintRect = renderer.fastRepaintRectInLocalCoordinates();
```
This branching looks inconsistent with other places where we pass the `context`. How difficult is it to merge `fastRepaintRectInLocalCoordinates()` in `repaintRectInLocalCoordinates()` and pass the `context` or another flag to it? 
```
enum class RepaintRectCalculation { Fast, Accurate };
virtual FloatRect repaintRectInLocalCoordinates(RepaintRectCalculation = RepaintRectCalculation::Fast) const;
```
In this case all the changes form calling `repaintRectInLocalCoordinates()` to `fastRepaintRectInLocalCoordinates()` in this patch are not needed.

This call will look like this:
```
auto repaintRect = renderer.repaintRectInLocalCoordinates(context.options.contains(RenderObject::VisibleRectContextOption::CalculateFastRepaintRect) ? RepaintRectCalculation::Fast : RepaintRectCalculation::Accurate);
```

And you can even add a member RepaintRectCalculation to VisibleRectContext instead of it as an option. So this call will look like this:

```
auto repaintRect = renderer.repaintRectInLocalCoordinates(context.calculation);
```

Constellation replied:
That sounds good! Changed.

shallawa commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.cpp:
```diff
@@ -362,7 +362,12 @@ LayoutRect LegacyRenderSVGRoot::clippedOverflowRect(const RenderLayerModelObject
+    if (context.options.contains(VisibleRectContextOption::DoNotUseSVGFastRepaintRectInLocalCoordinates))
+        repaintRectInLocalCoordinates = this->repaintRectInLocalCoordinates();
+    else
+        repaintRectInLocalCoordinates = this->fastRepaintRectInLocalCoordinates();
```
I think you are changing all the calls to `repaintRectInLocalCoordinates()` to `fastRepaintRectInLocalCoordinates()` unconditionally in all places expect this one, the one in `SVGRenderSupport::clippedOverflowRectForRepaint()` and the ones in `checkIntersection()` and `checkEnclosure()`

So I think it is worth looking at adding a calculation method flag to `repaintRectInLocalCoordinates()` and default it to `Fast` and removing `fastRepaintRectInLocalCoordinates()`.

Constellation replied:
Done.

shallawa commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -789,9 +787,10 @@ class RenderObject : public CachedImageClient, public CanMakeCheckedPtr {
         ApplyCompositedClips = 1 << 1,
         ApplyCompositedContainerScrolls  = 1 << 2,
         ApplyContainerClip = 1 << 3,
+        CalculateFastRepaintRect = 1 << 4,
```
I am sorry I did not search the code before making this suggestion. I think we should go for the opposite: `CalculateAccurateRepaintRect`. This will make less changes and it makes sense because it seems most of the callers of the VisibleRectContextOption need fast `repaintRect`.

Constellation replied:
Changed.

shallawa commented on Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
```diff
@@ -55,16 +55,16 @@
     // map to parent coords and recurse up the parent chain.
-    FloatRect repaintRect = renderer.repaintRectInLocalCoordinates();
-    return enclosingLayoutRect(renderer.computeFloatRectForRepaint(repaintRect, repaintContainer));
+    auto repaintRect = renderer.repaintRectInLocalCoordinates(context.options.contains(RenderObject::VisibleRectContextOption::CalculateFastRepaintRect) ? RenderObject::RepaintRectCalculation::Fast : RenderObject::RepaintRectCalculation::Accurate);
```
I think you can add an inline method in `VisibleRectContext`  to return `RepaintRectCalculation`. So this code can look like this:
```
auto repaintRect = renderer.repaintRectInLocalCoordinates(context.repaintRectCalculation());
```

Constellation replied:
Changed.

shallawa commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.cpp:
```diff
@@ -362,7 +362,8 @@ LayoutRect LegacyRenderSVGRoot::clippedOverflowRect(const RenderLayerModelObject
 
-    FloatRect contentRepaintRect = m_localToBorderBoxTransform.mapRect(repaintRectInLocalCoordinates());
+    auto repaintRectInLocalCoordinates = this->repaintRectInLocalCoordinates(context.options.contains(VisibleRectContextOption::CalculateFastRepaintRect) ? RenderObject::RepaintRectCalculation::Fast : RenderObject::RepaintRectCalculation::Accurate);
+    auto contentRepaintRect = m_localToBorderBoxTransform.mapRect(WTFMove(repaintRectInLocalCoordinates));
```
Is it faster to use `WTFMove()` here?

shallawa replied:
Also it be one statement like before:
```
auto contentRepaintRect = m_localToBorderBoxTransform.mapRect(repaintRectInLocalCoordinates(context.repaintRectCalculation()));
```

Constellation replied:
Changed.

Constellation committed:
[SVG] Add fast mode to repaintRectInLocalCoordinates
https://bugs.webkit.org/show_bug.cgi?id=262672
rdar://116499404

Reviewed by Said Abou-Hallawa.

This patch replaces repaintRectInLocalCoordinates calls with fast approximation version stubs. They are no difference at this point,
but approximation version will get implementation in a subsequent change.

1. This patch adds Fast and Accurate mode to repaintRectInLocalCoordinates. And replaces fastRepaintRectInLocalCoordinates with repaintRectInLocalCoordinates Fast mode.
2. We use repaintRectInLocalCoordinates Accurate mode in checkEnclosure / checkIntersection to keep the current behavior as is even after introducing approximation.
2. We add visibleRectContextForRenderTreeAsText to continue using repaintRectInLocalCoordinates Accurate mode for SVG RenderTreeAsText. This is having an option
   in VisibleRectContext and switch the calls based on this option.

* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::absoluteClippedOverflowRectForRenderTreeAsText const):
(WebCore::RenderObject::visibleRectContextForRenderTreeAsText):
* Source/WebCore/rendering/svg/RenderSVGBlock.cpp:
(WebCore::RenderSVGBlock::clippedOverflowRect const):
* Source/WebCore/rendering/svg/RenderSVGInline.cpp:
(WebCore::RenderSVGInline::fastRepaintRectInLocalCoordinates const):
(WebCore::RenderSVGInline::clippedOverflowRect const):
* Source/WebCore/rendering/svg/RenderSVGInline.h:
* Source/WebCore/rendering/svg/RenderSVGModelObject.cpp:
(WebCore::RenderSVGModelObject::addFocusRingRects const):
(WebCore::RenderSVGModelObject::checkIntersection):
(WebCore::RenderSVGModelObject::checkEnclosure):
* Source/WebCore/rendering/svg/RenderSVGResourceGradient.cpp:
(WebCore::createMaskAndSwapContextForTextGradient):
(WebCore::clipToTextMask):
* Source/WebCore/rendering/svg/RenderSVGResourceMarker.cpp:
(WebCore::RenderSVGResourceMarker::markerBoundaries const):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
(WebCore::RenderSVGResourceMasker::applyResource):
(WebCore::RenderSVGResourceMasker::calculateMaskContentRepaintRect):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::clippedOverflowRectForRepaint):
(WebCore::SVGRenderSupport::computeContainerBoundingBoxes):
* Source/WebCore/rendering/svg/SVGRenderSupport.h:
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
* Source/WebCore/rendering/svg/SVGRenderingContext.cpp:
(WebCore::SVGRenderingContext::prepareToRenderSVGContent):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.cpp:
(WebCore::LegacyRenderSVGContainer::addFocusRingRects const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.cpp:
(WebCore::LegacyRenderSVGImage::paint):
(WebCore::LegacyRenderSVGImage::addFocusRingRects const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.cpp:
(WebCore::LegacyRenderSVGModelObject::clippedOverflowRect const):
(WebCore::LegacyRenderSVGModelObject::outlineBoundsForRepaint const):
(WebCore::LegacyRenderSVGModelObject::absoluteFocusRingQuads):
(WebCore::LegacyRenderSVGModelObject::checkIntersection):
(WebCore::LegacyRenderSVGModelObject::checkEnclosure):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.cpp:
(WebCore::LegacyRenderSVGResourceClipper::applyResource):
(WebCore::LegacyRenderSVGResourceClipper::calculateClipContentRepaintRect):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::layout):
(WebCore::LegacyRenderSVGRoot::clippedOverflowRect const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::paint):
(WebCore::LegacyRenderSVGShape::addFocusRingRects const):
* Source/WebCore/svg/SVGFEImageElement.cpp:
(WebCore::SVGFEImageElement::imageBufferForEffect const):

Canonical link: https://commits.webkit.org/268940@main
